### PR TITLE
Change the default thread model for RR

### PR DESCRIPTION
**Pull request by stuartwdouglas**
- If an async type is returned (CompletionStage, Uni, Multi etc) then it
  remains non blocking
- If the method is a Kotlin suspended method it remains non-blocking
- All other methods default to blocking, unless explicitly specified
- @(Non)Blocking on an application class works as normal, and will
  override the method signature based defaults

### CODE REVIEW

File: independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/BlockingDefault.java
Comment (geoand) [inline independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/BlockingDefault.java:3]: Do we really want this to be able to change this?

Comment (FroMage) [reply independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/BlockingDefault.java:3]: Not sure I understand your question :(

**Review by FroMage — CHANGES_REQUESTED** [review]
LGTM except this must be mentioned in the docs, and perhaps as a breaking change too? Although in theory this shouldn't break any application unless they explicitly rely on specific threads.

File: independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/BlockingDefault.java
Comment (geoand) [reply independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/BlockingDefault.java:3]: IIUC, this allows a user to set what the default handling should be, no?

**Comment by stuartwdouglas** [comment]
I have added the docs (I was waiting for the initial feedback), and moved this out of draft.

**Comment by stuartwdouglas** [comment]
@FroMage I have updated the docs and the native test that failed to download does not look like it has anything to do with RR.

**Comment by stuartwdouglas** [comment]
@FroMage it would be good to get this in, so we are not trying to rush it at the end of the release cycle.

File: docs/src/main/asciidoc/resteasy-reactive.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/resteasy-reactive.adoc:966]:
```suggestion
By default, the thread RESTEasy Reactive will run endpoint methods on depends on the signature of the method.
```

**Review by FroMage — APPROVED** [review]
LGTM except the typo, can I let you fix it before you merge?

**Comment by stuartwdouglas** [comment]
That gradle windows test has failed 3 times, but as far as I can see resteasy-reactive is not used at all, so I don't see how it could be related

**Comment by stuartwdouglas** [comment]
@geoand are you happy with this? If so it should be good to go.

**Comment by geoand** [comment]
Yeah, let's get it in ASAP

**Review by geoand — APPROVED** [review]

**Comment by gsmet** [comment]
@bcluap would you have the time to check `main` after this commit with your application(s)?

**Comment by geoand** [comment]
We would also want @Sgitario and @rsvoboda to do some tests if they have any RESTEasy Reactive applications in their testsuites

**Comment by bcluap** [comment]
Yes sure. Let me know when

**Comment by geoand** [comment]
The change has been merged, so whenever you can test, we would be grateful

**Comment by bcluap** [comment]
Any version after 2.0.3.Final gives me this error so I cant run my tests:
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running guru.jini.arch.conductor.ConductorTest
11:15:43,729: [build-26]Failed to index org.eclipse.microprofile.health.Startup: Class does not exist in ClassLoader QuarkusClassLoader:Deployment Class Loader: TEST@69b08dc2
11:15:43,730: [build-26]Class for name: org.eclipse.microprofile.health.Startup was not found in Jandex index. Please ensure the class is part of the index.
[ERROR] Tests run: 13, Failures: 0, Errors: 1, Skipped: 12, Time elapsed: 5.262 s <<< FAILURE! - in guru.jini.arch.conductor.ConductorTest
[ERROR] guru.jini.arch.conductor.ConductorTest.initialiseConfigForTests  Time elapsed: 0.004 s  <<< ERROR!
java.lang.RuntimeException: 
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.NullPointerException: Annotation class not available: @Startup
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at io.quarkus.arc.processor.AnnotationLiteralProcessor.process(AnnotationLiteralProcessor.java:61)
	at io.quarkus.arc.processor.BeanGenerator.collectInjectionPointAnnotations(BeanGenerator.java:1888)
	at io.quarkus.arc.processor.BuiltinBean.lambda$static$0(BuiltinBean.java:40)
	at io.quarkus.arc.processor.BeanGenerator.initConstructor(BeanGenerator.java:663)
	at io.quarkus.arc.processor.BeanGenerator.createConstructor(BeanGenerator.java:585)
	at io.quarkus.arc.processor.BeanGenerator.generateClassBean(BeanGenerator.java:303)
	at io.quarkus.arc.processor.BeanGenerator.generate(BeanGenerator.java:116)
	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:189)
	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:457)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
	at java.base/java.lang.Thread.run(Thread.java:831)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

Caused by: java.lang.RuntimeException: 
io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.NullPointerException: Annotation class not available: @Startup
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at io.quarkus.arc.processor.AnnotationLiteralProcessor.process(AnnotationLiteralProcessor.java:61)
	at io.quarkus.arc.processor.BeanGenerator.collectInjectionPointAnnotations(BeanGenerator.java:1888)
	at io.quarkus.arc.processor.BuiltinBean.lambda$static$0(BuiltinBean.java:40)
	at io.quarkus.arc.processor.BeanGenerator.initConstructor(BeanGenerator.java:663)
	at io.quarkus.arc.processor.BeanGenerator.createConstructor(BeanGenerator.java:585)
	at io.quarkus.arc.processor.BeanGenerator.generateClassBean(BeanGenerator.java:303)
	at io.quarkus.arc.processor.BeanGenerator.generate(BeanGenerator.java:116)
	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:189)
	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:457)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
	at java.base/java.lang.Thread.run(Thread.java:831)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

Caused by: io.quarkus.builder.BuildException: 
Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.NullPointerException: Annotation class not available: @Startup
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at io.quarkus.arc.processor.AnnotationLiteralProcessor.process(AnnotationLiteralProcessor.java:61)
	at io.quarkus.arc.processor.BeanGenerator.collectInjectionPointAnnotations(BeanGenerator.java:1888)
	at io.quarkus.arc.processor.BuiltinBean.lambda$static$0(BuiltinBean.java:40)
	at io.quarkus.arc.processor.BeanGenerator.initConstructor(BeanGenerator.java:663)
	at io.quarkus.arc.processor.BeanGenerator.createConstructor(BeanGenerator.java:585)
	at io.quarkus.arc.processor.BeanGenerator.generateClassBean(BeanGenerator.java:303)
	at io.quarkus.arc.processor.BeanGenerator.generate(BeanGenerator.java:116)
	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:189)
	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:457)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
	at java.base/java.lang.Thread.run(Thread.java:831)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

Caused by: java.lang.NullPointerException: Annotation class not available: @Startup

```

**Comment by bcluap** [comment]
Using main the error looks like this:

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running guru.jini.arch.conductor.ConductorTest
[ERROR] Tests run: 13, Failures: 0, Errors: 1, Skipped: 12, Time elapsed: 4.228 s <<< FAILURE! - in guru.jini.arch.conductor.ConductorTest
[ERROR] guru.jini.arch.conductor.ConductorTest.initialiseConfigForTests  Time elapsed: 0.004 s  <<< ERROR!
java.lang.RuntimeException: java.lang.NoClassDefFoundError: org/eclipse/microprofile/health/Startup
Caused by: java.lang.NoClassDefFoundError: org/eclipse/microprofile/health/Startup
Caused by: java.lang.ClassNotFoundException: org.eclipse.microprofile.health.Startup
```

**Comment by geoand** [comment]
@mkouba any idea about ^

**Comment by bcluap** [comment]
FYI. My extensions in use:
11:28:51,966: [main]Installed features: [agroal, cdi, hibernate-orm, hibernate-validator, jaeger, jdbc-mysql, jdbc-postgresql, narayana-jta, resteasy-reactive, resteasy-reactive-jackson, smallrye-context-propagation, smallrye-health, smallrye-metrics, smallrye-openapi, smallrye-opentracing, swagger-ui, vertx]

**Comment by mkouba** [comment]
It seems that `org.eclipse.microprofile.health.Startup` is a qualifier that was added in MP Health 3.1 but apparently it's not available during build (i.e. it's not indexed by the extension and ArC is not able to index it lazily). Maybe some mixed dependencies problem? CC @xstefank.

**Comment by bcluap** [comment]
Thanks. We had a dependency on MP Health 3.0.

I changed to 3.1 and now its happier. Will run full tests now

**Comment by bcluap** [comment]
image
So far so good on some basic tests. Running full suit now

**Comment by bcluap** [comment]
Our full testsuite passed. Great stuff

**Comment by Sgitario** [comment]
> We would also want @Sgitario and @rsvoboda to do some tests if they have any RESTEasy Reactive applications in their testsuites

Thanks for letting us know! This is now under our radar.

**Pull request merged**
